In [69]:
import numpy as np
import pandas as pd
import seaborn as sns
pd.set_option("display.max_columns", None)

### Exploring basic data properties

In [143]:
housedf_raw = pd.read_csv('./data/Ames_HousePrice.csv', index_col=0)

In [144]:
housedf_raw.head()

PID  GrLivArea  SalePrice  MSSubClass MSZoning  LotFrontage  LotArea  \
1  909176150        856     126000          30       RL          NaN     7890   
2  905476230       1049     139500         120       RL         42.0     4235   
3  911128020       1001     124900          30  C (all)         60.0     6060   
4  535377150       1039     114000          70       RL         80.0     8146   
5  534177230       1665     227000          60       RL         70.0     8400   

  Street Alley LotShape LandContour Utilities LotConfig LandSlope  \
1   Pave   NaN      Reg         Lvl    AllPub    Corner       Gtl   
2   Pave   NaN      Reg         Lvl    AllPub    Inside       Gtl   
3   Pave   NaN      Reg         Lvl    AllPub    Inside       Gtl   
4   Pave   NaN      Reg         Lvl    AllPub    Corner       Gtl   
5   Pave   NaN      Reg         Lvl    AllPub    Inside       Gtl   

  Neighborhood Condition1 Condition2 BldgType HouseStyle  OverallQual  \
1        SWISU       Norm       Norm     1Fam     1Story            6   
2      Edwards       Norm       Norm   TwnhsE     1Story            5   
3       IDOTRR       Norm       Norm     1Fam     1Story            5   
4      OldTown       Norm       Norm     1Fam     2Story            4   
5       NWAmes       Norm       Norm     1Fam     2Story            8   

   OverallCond  YearBuilt  YearRemodAdd RoofStyle RoofMatl Exterior1st  \
1            6       1939          1950     Gable  CompShg     Wd Sdng   
2            5       1984          1984     Gable  CompShg     HdBoard   
3            9       1930          2007       Hip  CompShg     MetalSd   
4            8       1900          2003     Gable  CompShg     MetalSd   
5            6       2001          2001     Gable  CompShg     VinylSd   

  Exterior2nd MasVnrType  MasVnrArea ExterQual ExterCond Foundation BsmtQual  \
1     Wd Sdng       None         0.0        TA        TA     CBlock       TA   
2     HdBoard    BrkFace       149.0        Gd        TA     CBlock       Gd   
3     MetalSd       None         0.0        Gd        TA     BrkTil       TA   
4     MetalSd       None         0.0        Gd        Gd     BrkTil       Fa   
5     VinylSd       None         0.0        Gd        TA      PConc       Gd   

  BsmtCond BsmtExposure BsmtFinType1  BsmtFinSF1 BsmtFinType2  BsmtFinSF2  \
1       TA           No          Rec       238.0          Unf         0.0   
2       TA           Mn          GLQ       552.0          ALQ       393.0   
3       TA           No          ALQ       737.0          Unf         0.0   
4       TA           No          Unf         0.0          Unf         0.0   
5       TA           No          GLQ       643.0          Unf         0.0   

   BsmtUnfSF  TotalBsmtSF Heating HeatingQC CentralAir Electrical  1stFlrSF  \
1      618.0        856.0    GasA        TA          Y      SBrkr       856   
2      104.0       1049.0    GasA        TA          Y      SBrkr      1049   
3      100.0        837.0    GasA        Ex          Y      SBrkr      1001   
4      405.0        405.0    GasA        Gd          Y      SBrkr       717   
5      167.0        810.0    GasA        Ex          Y      SBrkr       810   

   2ndFlrSF  LowQualFinSF  BsmtFullBath  BsmtHalfBath  FullBath  HalfBath  \
1         0             0           1.0           0.0         1         0   
2         0             0           1.0           0.0         2         0   
3         0             0           0.0           0.0         1         0   
4       322             0           0.0           0.0         1         0   
5       855             0           1.0           0.0         2         1   

   BedroomAbvGr  KitchenAbvGr KitchenQual  TotRmsAbvGrd Functional  \
1             2             1          TA             4        Typ   
2             2             1          Gd             5        Typ   
3             2             1          Gd             5        Typ   
4             2             1          TA             6      

In [145]:
# convenient display of datatypes
pd.DataFrame(housedf_raw.dtypes).T

PID GrLivArea SalePrice MSSubClass MSZoning LotFrontage LotArea  Street  \
0  int64     int64     int64      int64   object     float64   int64  object   

    Alley LotShape LandContour Utilities LotConfig LandSlope Neighborhood  \
0  object   object      object    object    object    object       object   

  Condition1 Condition2 BldgType HouseStyle OverallQual OverallCond YearBuilt  \
0     object     object   object     object       int64       int64     int64   

  YearRemodAdd RoofStyle RoofMatl Exterior1st Exterior2nd MasVnrType  \
0        int64    object   object      object      object     object   

  MasVnrArea ExterQual ExterCond Foundation BsmtQual BsmtCond BsmtExposure  \
0    float64    object    object     object   object   object       object   

  BsmtFinType1 BsmtFinSF1 BsmtFinType2 BsmtFinSF2 BsmtUnfSF TotalBsmtSF  \
0       object    float64       object    float64   float64     float64   

  Heating HeatingQC CentralAir Electrical 1stFlrSF 2ndFlrSF LowQualFinSF  \
0  object    object     object     object    int64    int64        int64   

  BsmtFullBath BsmtHalfBath FullBath HalfBath BedroomAbvGr KitchenAbvGr  \
0      float64      float64    int64    int64        int64        int64   

  KitchenQual TotRmsAbvGrd Functional Fireplaces FireplaceQu GarageType  \
0      object        int64     object      int64      object     object   

  GarageYrBlt GarageFinish GarageCars GarageArea GarageQual GarageCond  \
0     float64       object    float64    float64     object     object   

  PavedDrive WoodDeckSF OpenPorchSF EnclosedPorch 3SsnPorch ScreenPorch  \
0     object      int64       int64         int64     int64       int64   

  PoolArea  PoolQC   Fence MiscFeature MiscVal MoSold YrSold SaleType  \
0    int64  object  object      object   int64  int64  int64   object   

  SaleCondition  
0        object

Note: we will need to convert all of the string columns to categorical features later

In [146]:
housedf = housedf_raw.copy()
housedf.head()

PID  GrLivArea  SalePrice  MSSubClass MSZoning  LotFrontage  LotArea  \
1  909176150        856     126000          30       RL          NaN     7890   
2  905476230       1049     139500         120       RL         42.0     4235   
3  911128020       1001     124900          30  C (all)         60.0     6060   
4  535377150       1039     114000          70       RL         80.0     8146   
5  534177230       1665     227000          60       RL         70.0     8400   

  Street Alley LotShape LandContour Utilities LotConfig LandSlope  \
1   Pave   NaN      Reg         Lvl    AllPub    Corner       Gtl   
2   Pave   NaN      Reg         Lvl    AllPub    Inside       Gtl   
3   Pave   NaN      Reg         Lvl    AllPub    Inside       Gtl   
4   Pave   NaN      Reg         Lvl    AllPub    Corner       Gtl   
5   Pave   NaN      Reg         Lvl    AllPub    Inside       Gtl   

  Neighborhood Condition1 Condition2 BldgType HouseStyle  OverallQual  \
1        SWISU       Norm       Norm     1Fam     1Story            6   
2      Edwards       Norm       Norm   TwnhsE     1Story            5   
3       IDOTRR       Norm       Norm     1Fam     1Story            5   
4      OldTown       Norm       Norm     1Fam     2Story            4   
5       NWAmes       Norm       Norm     1Fam     2Story            8   

   OverallCond  YearBuilt  YearRemodAdd RoofStyle RoofMatl Exterior1st  \
1            6       1939          1950     Gable  CompShg     Wd Sdng   
2            5       1984          1984     Gable  CompShg     HdBoard   
3            9       1930          2007       Hip  CompShg     MetalSd   
4            8       1900          2003     Gable  CompShg     MetalSd   
5            6       2001          2001     Gable  CompShg     VinylSd   

  Exterior2nd MasVnrType  MasVnrArea ExterQual ExterCond Foundation BsmtQual  \
1     Wd Sdng       None         0.0        TA        TA     CBlock       TA   
2     HdBoard    BrkFace       149.0        Gd        TA     CBlock       Gd   
3     MetalSd       None         0.0        Gd        TA     BrkTil       TA   
4     MetalSd       None         0.0        Gd        Gd     BrkTil       Fa   
5     VinylSd       None         0.0        Gd        TA      PConc       Gd   

  BsmtCond BsmtExposure BsmtFinType1  BsmtFinSF1 BsmtFinType2  BsmtFinSF2  \
1       TA           No          Rec       238.0          Unf         0.0   
2       TA           Mn          GLQ       552.0          ALQ       393.0   
3       TA           No          ALQ       737.0          Unf         0.0   
4       TA           No          Unf         0.0          Unf         0.0   
5       TA           No          GLQ       643.0          Unf         0.0   

   BsmtUnfSF  TotalBsmtSF Heating HeatingQC CentralAir Electrical  1stFlrSF  \
1      618.0        856.0    GasA        TA          Y      SBrkr       856   
2      104.0       1049.0    GasA        TA          Y      SBrkr      1049   
3      100.0        837.0    GasA        Ex          Y      SBrkr      1001   
4      405.0        405.0    GasA        Gd          Y      SBrkr       717   
5      167.0        810.0    GasA        Ex          Y      SBrkr       810   

   2ndFlrSF  LowQualFinSF  BsmtFullBath  BsmtHalfBath  FullBath  HalfBath  \
1         0             0           1.0           0.0         1         0   
2         0             0           1.0           0.0         2         0   
3         0             0           0.0           0.0         1         0   
4       322             0           0.0           0.0         1         0   
5       855             0           1.0           0.0         2         1   

   BedroomAbvGr  KitchenAbvGr KitchenQual  TotRmsAbvGrd Functional  \
1             2             1          TA             4        Typ   
2             2             1          Gd             5        Typ   
3             2             1          Gd             5        Typ   
4             2             1          TA             6      

In [147]:
housedf.shape
# you should see 2580 samples with 81 features

(2580, 81)

### Examining null values

In [148]:
# counts of null values per column
null_counts = pd.DataFrame(pd.isnull(housedf).sum(axis=0)).T
null_counts

PID  GrLivArea  SalePrice  MSSubClass  MSZoning  LotFrontage  LotArea  \
0    0          0          0           0         0          462        0   

   Street  Alley  LotShape  LandContour  Utilities  LotConfig  LandSlope  \
0       0   2412         0            0          0          0          0   

   Neighborhood  Condition1  Condition2  BldgType  HouseStyle  OverallQual  \
0             0           0           0         0           0            0   

   OverallCond  YearBuilt  YearRemodAdd  RoofStyle  RoofMatl  Exterior1st  \
0            0          0             0          0         0            0   

   Exterior2nd  MasVnrType  MasVnrArea  ExterQual  ExterCond  Foundation  \
0            0          14          14          0          0           0   

   BsmtQual  BsmtCond  BsmtExposure  BsmtFinType1  BsmtFinSF1  BsmtFinType2  \
0        69        69            71            69           1            70   

   BsmtFinSF2  BsmtUnfSF  TotalBsmtSF  Heating  HeatingQC  CentralAir  \
0           1          1            1        0          0           0   

   Electrical  1stFlrSF  2ndFlrSF  LowQualFinSF  BsmtFullBath  BsmtHalfBath  \
0           1         0         0             0             2             2   

   FullBath  HalfBath  BedroomAbvGr  KitchenAbvGr  KitchenQual  TotRmsAbvGrd  \
0         0         0             0             0            0             0   

   Functional  Fireplaces  FireplaceQu  GarageType  GarageYrBlt  GarageFinish  \
0           0           0         1241         127          129           129   

   GarageCars  GarageArea  GarageQual  GarageCond  PavedDrive  WoodDeckSF  \
0           1           1         129         129           0           0   

   OpenPorchSF  EnclosedPorch  3SsnPorch  ScreenPorch  PoolArea  PoolQC  \
0            0              0          0            0         0    2571   

   Fence  MiscFeature  MiscVal  MoSold  YrSold  SaleType  SaleCondition  
0   2055         2483        0       0       0         0              0

In [149]:
# get a list with names of columns containing null values
cols_with_null = [col for col in housedf.columns if np.sum(pd.isnull(housedf[col])) != 0]

In [150]:
# display a sample with the respective feature missing for each feature containing null value
for col in cols_with_null:
    print(col)
    display(housedf.loc[pd.isnull(housedf[col]),:].sample(1))

LotFrontage


PID  GrLivArea  SalePrice  MSSubClass MSZoning  LotFrontage  \
761  914465060       1689     271000          20       RL          NaN   

     LotArea Street Alley LotShape LandContour Utilities LotConfig LandSlope  \
761    13265   Pave   NaN      IR1         Lvl    AllPub   CulDSac       Gtl   

    Neighborhood Condition1 Condition2 BldgType HouseStyle  OverallQual  \
761      Mitchel       Norm       Norm     1Fam     1Story            8   

     OverallCond  YearBuilt  YearRemodAdd RoofStyle RoofMatl Exterior1st  \
761            5       2002          2002       Hip  CompShg     CemntBd   

    Exterior2nd MasVnrType  MasVnrArea ExterQual ExterCond Foundation  \
761     CmentBd    BrkFace       148.0        Gd        TA      PConc   

    BsmtQual BsmtCond BsmtExposure BsmtFinType1  BsmtFinSF1 BsmtFinType2  \
761       Gd       TA           No          GLQ      1218.0          Unf   

     BsmtFinSF2  BsmtUnfSF  TotalBsmtSF Heating HeatingQC CentralAir  \
761         0.0      350.0       1568.0    GasA        Ex          Y   

    Electrical  1stFlrSF  2ndFlrSF  LowQualFinSF  BsmtFullBath  BsmtHalfBath  \
761      SBrkr      1689         0             0           1.0           0.0   

     FullBath  HalfBath  BedroomAbvGr  KitchenAbvGr KitchenQual  TotRmsAbvGrd  \
761         2         0             3             1          Gd             7   

    Functional  Fireplaces FireplaceQu GarageType  GarageYrBlt GarageFinish  \
761        Typ           2          Gd     Attchd       2002.0          RFn   

     GarageCars  GarageArea GarageQual GarageCond PavedDrive  WoodDeckSF  \
761         3.0       857.0         TA         TA          Y         150   

     OpenPorchSF  EnclosedPorch  3SsnPorch  ScreenPorch  PoolArea PoolQC  \
761           59              0          0            0         0    NaN   

    Fence MiscFeature  MiscVal  MoSold  YrSold SaleType SaleCondition  
761   NaN         NaN        0       7    2008      WD         Normal

Alley


PID  GrLivArea  SalePrice  MSSubClass MSZoning  LotFrontage  \
909  534427040       1252     142000          20       RL         80.0   

     LotArea Street Alley LotShape LandContour Utilities LotConfig LandSlope  \
909    10721   Pave   NaN      IR1         Lvl    AllPub    Inside       Gtl   

    Neighborhood Condition1 Condition2 BldgType HouseStyle  OverallQual  \
909        NAmes       Norm       Norm     1Fam     1Story            6   

     OverallCond  YearBuilt  YearRemodAdd RoofStyle RoofMatl Exterior1st  \
909            6       1959          1959       Hip  CompShg     HdBoard   

    Exterior2nd MasVnrType  MasVnrArea ExterQual ExterCond Foundation  \
909     HdBoard      Stone       243.0        Gd        TA     CBlock   

    BsmtQual BsmtCond BsmtExposure BsmtFinType1  BsmtFinSF1 BsmtFinType2  \
909       TA       TA           No          Unf         0.0          Unf   

     BsmtFinSF2  BsmtUnfSF  TotalBsmtSF Heating HeatingQC CentralAir  \
909         0.0     1252.0       1252.0    GasA        Ex          Y   

    Electrical  1stFlrSF  2ndFlrSF  LowQualFinSF  BsmtFullBath  BsmtHalfBath  \
909      SBrkr      1252         0             0           0.0           0.0   

     FullBath  HalfBath  BedroomAbvGr  KitchenAbvGr KitchenQual  TotRmsAbvGrd  \
909         1         0             3             1          Gd             7   

    Functional  Fireplaces FireplaceQu GarageType  GarageYrBlt GarageFinish  \
909        Typ           0         NaN     Detchd       1960.0          Unf   

     GarageCars  GarageArea GarageQual GarageCond PavedDrive  WoodDeckSF  \
909         2.0       528.0         TA         TA          Y           0   

     OpenPorchSF  EnclosedPorch  3SsnPorch  ScreenPorch  PoolArea PoolQC  \
909           39              0          0            0         0    NaN   

    Fence MiscFeature  MiscVal  MoSold  YrSold SaleType SaleCondition  
909   NaN         NaN        0      10    2008      WD         Normal

MasVnrType


PID  GrLivArea  SalePrice  MSSubClass MSZoning  LotFrontage  \
892  528290090       1414     176000          60       RL          NaN   

     LotArea Street Alley LotShape LandContour Utilities LotConfig LandSlope  \
892     7750   Pave   NaN      Reg         Lvl    AllPub    Inside       Gtl   

    Neighborhood Condition1 Condition2 BldgType HouseStyle  OverallQual  \
892      Gilbert       RRAn       Norm     1Fam     2Story            7   

     OverallCond  YearBuilt  YearRemodAdd RoofStyle RoofMatl Exterior1st  \
892            5       2002          2002     Gable  CompShg     VinylSd   

    Exterior2nd MasVnrType  MasVnrArea ExterQual ExterCond Foundation  \
892     VinylSd        NaN         NaN        Gd        TA      PConc   

    BsmtQual BsmtCond BsmtExposure BsmtFinType1  BsmtFinSF1 BsmtFinType2  \
892       Gd       TA           No          Unf         0.0          Unf   

     BsmtFinSF2  BsmtUnfSF  TotalBsmtSF Heating HeatingQC CentralAir  \
892         0.0      707.0        707.0    GasA        Ex          Y   

    Electrical  1stFlrSF  2ndFlrSF  LowQualFinSF  BsmtFullBath  BsmtHalfBath  \
892      SBrkr       707       707             0           0.0           0.0   

     FullBath  HalfBath  BedroomAbvGr  KitchenAbvGr KitchenQual  TotRmsAbvGrd  \
892         2         1             3             1          Gd             6   

    Functional  Fireplaces FireplaceQu GarageType  GarageYrBlt GarageFinish  \
892        Typ           1          Gd     Attchd       2002.0          Fin   

     GarageCars  GarageArea GarageQual GarageCond PavedDrive  WoodDeckSF  \
892         2.0       403.0         TA         TA          Y         100   

     OpenPorchSF  EnclosedPorch  3SsnPorch  ScreenPorch  PoolArea PoolQC  \
892           35              0          0            0         0    NaN   

    Fence MiscFeature  MiscVal  MoSold  YrSold SaleType SaleCondition  
892   NaN         NaN        0       4    2008      WD         Normal

MasVnrArea


PID  GrLivArea  SalePrice  MSSubClass MSZoning  LotFrontage  \
356  528480160       1665     205950          60       FV         65.0   

     LotArea Street Alley LotShape LandContour Utilities LotConfig LandSlope  \
356     8125   Pave   NaN      Reg         Lvl    AllPub    Inside       Gtl   

    Neighborhood Condition1 Condition2 BldgType HouseStyle  OverallQual  \
356      Somerst       Norm       Norm     1Fam     2Story            7   

     OverallCond  YearBuilt  YearRemodAdd RoofStyle RoofMatl Exterior1st  \
356            6       2007          2007     Gable  CompShg     CemntBd   

    Exterior2nd MasVnrType  MasVnrArea ExterQual ExterCond Foundation  \
356     CmentBd        NaN         NaN        Gd        TA      PConc   

    BsmtQual BsmtCond BsmtExposure BsmtFinType1  BsmtFinSF1 BsmtFinType2  \
356       Gd       TA           No          Unf         0.0          Unf   

     BsmtFinSF2  BsmtUnfSF  TotalBsmtSF Heating HeatingQC CentralAir  \
356         0.0      813.0        813.0    GasA        Ex          Y   

    Electrical  1stFlrSF  2ndFlrSF  LowQualFinSF  BsmtFullBath  BsmtHalfBath  \
356      SBrkr       822       843             0           0.0           0.0   

     FullBath  HalfBath  BedroomAbvGr  KitchenAbvGr KitchenQual  TotRmsAbvGrd  \
356         2         1             3             1          Gd             7   

    Functional  Fireplaces FireplaceQu GarageType  GarageYrBlt GarageFinish  \
356        Typ           0         NaN     Attchd       2007.0          RFn   

     GarageCars  GarageArea GarageQual GarageCond PavedDrive  WoodDeckSF  \
356         2.0       562.0         TA         TA          Y           0   

     OpenPorchSF  EnclosedPorch  3SsnPorch  ScreenPorch  PoolArea PoolQC  \
356            0              0          0            0         0    NaN   

    Fence MiscFeature  MiscVal  MoSold  YrSold SaleType SaleCondition  
356   NaN         NaN        0       5    2008      WD         Normal

BsmtQual


PID  GrLivArea  SalePrice  MSSubClass MSZoning  LotFrontage  \
905  910206010        845      68104          50       RM         64.0   

     LotArea Street Alley LotShape LandContour Utilities LotConfig LandSlope  \
905    11067   Pave   NaN      Reg         Lvl    AllPub    Corner       Gtl   

    Neighborhood Condition1 Condition2 BldgType HouseStyle  OverallQual  \
905       IDOTRR       Norm       Norm     1Fam     1.5Fin            2   

     OverallCond  YearBuilt  YearRemodAdd RoofStyle RoofMatl Exterior1st  \
905            4       1939          1950     Gable  CompShg     MetalSd   

    Exterior2nd MasVnrType  MasVnrArea ExterQual ExterCond Foundation  \
905     MetalSd       None         0.0        TA        TA     BrkTil   

    BsmtQual BsmtCond BsmtExposure BsmtFinType1  BsmtFinSF1 BsmtFinType2  \
905      NaN      NaN          NaN          NaN         0.0          NaN   

     BsmtFinSF2  BsmtUnfSF  TotalBsmtSF Heating HeatingQC CentralAir  \
905         0.0        0.0          0.0    GasA        Ex          Y   

    Electrical  1stFlrSF  2ndFlrSF  LowQualFinSF  BsmtFullBath  BsmtHalfBath  \
905      SBrkr       640         0           205           0.0           0.0   

     FullBath  HalfBath  BedroomAbvGr  KitchenAbvGr KitchenQual  TotRmsAbvGrd  \
905         1         0             1             1          TA             4   

    Functional  Fireplaces FireplaceQu GarageType  GarageYrBlt GarageFinish  \
905       Maj2           0         NaN     Detchd       1950.0          Unf   

     GarageCars  GarageArea GarageQual GarageCond PavedDrive  WoodDeckSF  \
905         1.0       256.0         TA         Fa          N          48   

     OpenPorchSF  EnclosedPorch  3SsnPorch  ScreenPorch  PoolArea PoolQC  \
905            0              0          0            0         0    NaN   

    Fence MiscFeature  MiscVal  MoSold  YrSold SaleType SaleCondition  
905   NaN         NaN        0       3    2007      WD         Normal

BsmtCond


PID  GrLivArea  SalePrice  MSSubClass MSZoning  LotFrontage  \
368  535451210       1040     109900          20       RL         60.0   

     LotArea Street Alley LotShape LandContour Utilities LotConfig LandSlope  \
368     7200   Pave   NaN      Reg         Lvl    AllPub    Inside       Gtl   

    Neighborhood Condition1 Condition2 BldgType HouseStyle  OverallQual  \
368        NAmes       Norm       Norm     1Fam     1Story            4   

     OverallCond  YearBuilt  YearRemodAdd RoofStyle RoofMatl Exterior1st  \
368            5       1950          1950     Gable  CompShg     BrkFace   

    Exterior2nd MasVnrType  MasVnrArea ExterQual ExterCond Foundation  \
368     Wd Sdng       None         0.0        TA        TA       Slab   

    BsmtQual BsmtCond BsmtExposure BsmtFinType1  BsmtFinSF1 BsmtFinType2  \
368      NaN      NaN          NaN          NaN         0.0          NaN   

     BsmtFinSF2  BsmtUnfSF  TotalBsmtSF Heating HeatingQC CentralAir  \
368         0.0        0.0          0.0    GasA        TA          Y   

    Electrical  1stFlrSF  2ndFlrSF  LowQualFinSF  BsmtFullBath  BsmtHalfBath  \
368      FuseA      1040         0             0           0.0           0.0   

     FullBath  HalfBath  BedroomAbvGr  KitchenAbvGr KitchenQual  TotRmsAbvGrd  \
368         1         0             2             1          TA             4   

    Functional  Fireplaces FireplaceQu GarageType  GarageYrBlt GarageFinish  \
368        Typ           0         NaN     Detchd       1950.0          Unf   

     GarageCars  GarageArea GarageQual GarageCond PavedDrive  WoodDeckSF  \
368         2.0       420.0         TA         TA          Y           0   

     OpenPorchSF  EnclosedPorch  3SsnPorch  ScreenPorch  PoolArea PoolQC  \
368           29              0          0            0         0    NaN   

    Fence MiscFeature  MiscVal  MoSold  YrSold SaleType SaleCondition  
368   NaN         NaN        0       7    2006      WD         Normal

BsmtExposure


PID  GrLivArea  SalePrice  MSSubClass MSZoning  LotFrontage  \
295  534276190       1014      98000          20       RL         74.0   

     LotArea Street Alley LotShape LandContour Utilities LotConfig LandSlope  \
295     7785   Pave   NaN      IR1         Lvl    AllPub    Inside       Gtl   

    Neighborhood Condition1 Condition2 BldgType HouseStyle  OverallQual  \
295        NAmes       Norm       Norm     1Fam     1Story            5   

     OverallCond  YearBuilt  YearRemodAdd RoofStyle RoofMatl Exterior1st  \
295            5       1956          1956     Gable  CompShg     HdBoard   

    Exterior2nd MasVnrType  MasVnrArea ExterQual ExterCond Foundation  \
295     HdBoard       None         0.0        TA        TA       Slab   

    BsmtQual BsmtCond BsmtExposure BsmtFinType1  BsmtFinSF1 BsmtFinType2  \
295      NaN      NaN          NaN          NaN         0.0          NaN   

     BsmtFinSF2  BsmtUnfSF  TotalBsmtSF Heating HeatingQC CentralAir  \
295         0.0        0.0          0.0    GasA        TA          Y   

    Electrical  1stFlrSF  2ndFlrSF  LowQualFinSF  BsmtFullBath  BsmtHalfBath  \
295      SBrkr      1014         0             0           0.0           0.0   

     FullBath  HalfBath  BedroomAbvGr  KitchenAbvGr KitchenQual  TotRmsAbvGrd  \
295         1         0             2             1          TA             6   

    Functional  Fireplaces FireplaceQu GarageType  GarageYrBlt GarageFinish  \
295        Typ           0         NaN     Attchd       1956.0          RFn   

     GarageCars  GarageArea GarageQual GarageCond PavedDrive  WoodDeckSF  \
295         1.0       267.0         TA         TA          Y           0   

     OpenPorchSF  EnclosedPorch  3SsnPorch  ScreenPorch  PoolArea PoolQC  \
295            0             40          0          200         0    NaN   

    Fence MiscFeature  MiscVal  MoSold  YrSold SaleType SaleCondition  
295  GdWo         NaN        0       3    2008      WD         Normal

BsmtFinType1


PID  GrLivArea  SalePrice  MSSubClass MSZoning  LotFrontage  \
216  535450160       1040      92900          90       RL         60.0   

     LotArea Street Alley LotShape LandContour Utilities LotConfig LandSlope  \
216     8544   Pave   NaN      Reg         Lvl    AllPub    Corner       Gtl   

    Neighborhood Condition1 Condition2 BldgType HouseStyle  OverallQual  \
216        NAmes       Norm       Norm   Duplex     1Story            3   

     OverallCond  YearBuilt  YearRemodAdd RoofStyle RoofMatl Exterior1st  \
216            4       1950          1950     Gable  CompShg     BrkFace   

    Exterior2nd MasVnrType  MasVnrArea ExterQual ExterCond Foundation  \
216     BrkFace       None         0.0        TA        TA       Slab   

    BsmtQual BsmtCond BsmtExposure BsmtFinType1  BsmtFinSF1 BsmtFinType2  \
216      NaN      NaN          NaN          NaN         0.0          NaN   

     BsmtFinSF2  BsmtUnfSF  TotalBsmtSF Heating HeatingQC CentralAir  \
216         0.0        0.0          0.0    Wall        Fa          N   

    Electrical  1stFlrSF  2ndFlrSF  LowQualFinSF  BsmtFullBath  BsmtHalfBath  \
216      FuseA      1040         0             0           0.0           0.0   

     FullBath  HalfBath  BedroomAbvGr  KitchenAbvGr KitchenQual  TotRmsAbvGrd  \
216         2         0             2             2          TA             6   

    Functional  Fireplaces FireplaceQu GarageType  GarageYrBlt GarageFinish  \
216        Typ           0         NaN     Detchd       1987.0          Unf   

     GarageCars  GarageArea GarageQual GarageCond PavedDrive  WoodDeckSF  \
216         2.0       400.0         TA         TA          Y           0   

     OpenPorchSF  EnclosedPorch  3SsnPorch  ScreenPorch  PoolArea PoolQC  \
216            0              0          0            0         0    NaN   

    Fence MiscFeature  MiscVal  MoSold  YrSold SaleType SaleCondition  
216   NaN         NaN        0       6    2009      WD         Normal

BsmtFinSF1


PID  GrLivArea  SalePrice  MSSubClass MSZoning  LotFrontage  \
913  903230120        896      79000          20       RM         99.0   

     LotArea Street Alley LotShape LandContour Utilities LotConfig LandSlope  \
913     5940   Pave   NaN      IR1         Lvl    AllPub       FR3       Gtl   

    Neighborhood Condition1 Condition2 BldgType HouseStyle  OverallQual  \
913      BrkSide      Feedr       Norm     1Fam     1Story            4   

     OverallCond  YearBuilt  YearRemodAdd RoofStyle RoofMatl Exterior1st  \
913            7       1946          1950     Gable  CompShg     MetalSd   

    Exterior2nd MasVnrType  MasVnrArea ExterQual ExterCond Foundation  \
913      CBlock       None         0.0        TA        TA      PConc   

    BsmtQual BsmtCond BsmtExposure BsmtFinType1  BsmtFinSF1 BsmtFinType2  \
913      NaN      NaN          NaN          NaN         NaN          NaN   

     BsmtFinSF2  BsmtUnfSF  TotalBsmtSF Heating HeatingQC CentralAir  \
913         NaN        NaN          NaN    GasA        TA          Y   

    Electrical  1stFlrSF  2ndFlrSF  LowQualFinSF  BsmtFullBath  BsmtHalfBath  \
913      FuseA       896         0             0           NaN           NaN   

     FullBath  HalfBath  BedroomAbvGr  KitchenAbvGr KitchenQual  TotRmsAbvGrd  \
913         1         0             2             1          TA             4   

    Functional  Fireplaces FireplaceQu GarageType  GarageYrBlt GarageFinish  \
913        Typ           0         NaN     Detchd       1946.0          Unf   

     GarageCars  GarageArea GarageQual GarageCond PavedDrive  WoodDeckSF  \
913         1.0       280.0         TA         TA          Y           0   

     OpenPorchSF  EnclosedPorch  3SsnPorch  ScreenPorch  PoolArea PoolQC  \
913            0              0          0            0         0    NaN   

     Fence MiscFeature  MiscVal  MoSold  YrSold SaleType SaleCondition  
913  MnPrv         NaN        0       4    2008    ConLD       Abnorml

BsmtFinType2


PID  GrLivArea  SalePrice  MSSubClass MSZoning  LotFrontage  \
648  905477010       1124      93000          20       RL         77.0   

     LotArea Street Alley LotShape LandContour Utilities LotConfig LandSlope  \
648     8335   Pave   NaN      Reg         Lvl    AllPub    Corner       Gtl   

    Neighborhood Condition1 Condition2 BldgType HouseStyle  OverallQual  \
648      Edwards       Norm       Norm     1Fam     1Story            5   

     OverallCond  YearBuilt  YearRemodAdd RoofStyle RoofMatl Exterior1st  \
648            5       1954          1954     Gable  CompShg     Wd Sdng   

    Exterior2nd MasVnrType  MasVnrArea ExterQual ExterCond Foundation  \
648     Wd Sdng       None         0.0        TA        TA       Slab   

    BsmtQual BsmtCond BsmtExposure BsmtFinType1  BsmtFinSF1 BsmtFinType2  \
648      NaN      NaN          NaN          NaN         0.0          NaN   

     BsmtFinSF2  BsmtUnfSF  TotalBsmtSF Heating HeatingQC CentralAir  \
648         0.0        0.0          0.0    GasA        Gd          Y   

    Electrical  1stFlrSF  2ndFlrSF  LowQualFinSF  BsmtFullBath  BsmtHalfBath  \
648      SBrkr      1124         0             0           0.0           0.0   

     FullBath  HalfBath  BedroomAbvGr  KitchenAbvGr KitchenQual  TotRmsAbvGrd  \
648         1         0             3             1          TA             5   

    Functional  Fireplaces FireplaceQu GarageType  GarageYrBlt GarageFinish  \
648       Min2           1          Gd        NaN          NaN          NaN   

     GarageCars  GarageArea GarageQual GarageCond PavedDrive  WoodDeckSF  \
648         0.0         0.0        NaN        NaN          N           0   

     OpenPorchSF  EnclosedPorch  3SsnPorch  ScreenPorch  PoolArea PoolQC  \
648           36            190          0            0         0    NaN   

    Fence MiscFeature  MiscVal  MoSold  YrSold SaleType SaleCondition  
648   NaN         NaN        0       4    2006      WD         Normal

BsmtFinSF2


PID  GrLivArea  SalePrice  MSSubClass MSZoning  LotFrontage  \
913  903230120        896      79000          20       RM         99.0   

     LotArea Street Alley LotShape LandContour Utilities LotConfig LandSlope  \
913     5940   Pave   NaN      IR1         Lvl    AllPub       FR3       Gtl   

    Neighborhood Condition1 Condition2 BldgType HouseStyle  OverallQual  \
913      BrkSide      Feedr       Norm     1Fam     1Story            4   

     OverallCond  YearBuilt  YearRemodAdd RoofStyle RoofMatl Exterior1st  \
913            7       1946          1950     Gable  CompShg     MetalSd   

    Exterior2nd MasVnrType  MasVnrArea ExterQual ExterCond Foundation  \
913      CBlock       None         0.0        TA        TA      PConc   

    BsmtQual BsmtCond BsmtExposure BsmtFinType1  BsmtFinSF1 BsmtFinType2  \
913      NaN      NaN          NaN          NaN         NaN          NaN   

     BsmtFinSF2  BsmtUnfSF  TotalBsmtSF Heating HeatingQC CentralAir  \
913         NaN        NaN          NaN    GasA        TA          Y   

    Electrical  1stFlrSF  2ndFlrSF  LowQualFinSF  BsmtFullBath  BsmtHalfBath  \
913      FuseA       896         0             0           NaN           NaN   

     FullBath  HalfBath  BedroomAbvGr  KitchenAbvGr KitchenQual  TotRmsAbvGrd  \
913         1         0             2             1          TA             4   

    Functional  Fireplaces FireplaceQu GarageType  GarageYrBlt GarageFinish  \
913        Typ           0         NaN     Detchd       1946.0          Unf   

     GarageCars  GarageArea GarageQual GarageCond PavedDrive  WoodDeckSF  \
913         1.0       280.0         TA         TA          Y           0   

     OpenPorchSF  EnclosedPorch  3SsnPorch  ScreenPorch  PoolArea PoolQC  \
913            0              0          0            0         0    NaN   

     Fence MiscFeature  MiscVal  MoSold  YrSold SaleType SaleCondition  
913  MnPrv         NaN        0       4    2008    ConLD       Abnorml

BsmtUnfSF


PID  GrLivArea  SalePrice  MSSubClass MSZoning  LotFrontage  \
913  903230120        896      79000          20       RM         99.0   

     LotArea Street Alley LotShape LandContour Utilities LotConfig LandSlope  \
913     5940   Pave   NaN      IR1         Lvl    AllPub       FR3       Gtl   

    Neighborhood Condition1 Condition2 BldgType HouseStyle  OverallQual  \
913      BrkSide      Feedr       Norm     1Fam     1Story            4   

     OverallCond  YearBuilt  YearRemodAdd RoofStyle RoofMatl Exterior1st  \
913            7       1946          1950     Gable  CompShg     MetalSd   

    Exterior2nd MasVnrType  MasVnrArea ExterQual ExterCond Foundation  \
913      CBlock       None         0.0        TA        TA      PConc   

    BsmtQual BsmtCond BsmtExposure BsmtFinType1  BsmtFinSF1 BsmtFinType2  \
913      NaN      NaN          NaN          NaN         NaN          NaN   

     BsmtFinSF2  BsmtUnfSF  TotalBsmtSF Heating HeatingQC CentralAir  \
913         NaN        NaN          NaN    GasA        TA          Y   

    Electrical  1stFlrSF  2ndFlrSF  LowQualFinSF  BsmtFullBath  BsmtHalfBath  \
913      FuseA       896         0             0           NaN           NaN   

     FullBath  HalfBath  BedroomAbvGr  KitchenAbvGr KitchenQual  TotRmsAbvGrd  \
913         1         0             2             1          TA             4   

    Functional  Fireplaces FireplaceQu GarageType  GarageYrBlt GarageFinish  \
913        Typ           0         NaN     Detchd       1946.0          Unf   

     GarageCars  GarageArea GarageQual GarageCond PavedDrive  WoodDeckSF  \
913         1.0       280.0         TA         TA          Y           0   

     OpenPorchSF  EnclosedPorch  3SsnPorch  ScreenPorch  PoolArea PoolQC  \
913            0              0          0            0         0    NaN   

     Fence MiscFeature  MiscVal  MoSold  YrSold SaleType SaleCondition  
913  MnPrv         NaN        0       4    2008    ConLD       Abnorml

TotalBsmtSF


PID  GrLivArea  SalePrice  MSSubClass MSZoning  LotFrontage  \
913  903230120        896      79000          20       RM         99.0   

     LotArea Street Alley LotShape LandContour Utilities LotConfig LandSlope  \
913     5940   Pave   NaN      IR1         Lvl    AllPub       FR3       Gtl   

    Neighborhood Condition1 Condition2 BldgType HouseStyle  OverallQual  \
913      BrkSide      Feedr       Norm     1Fam     1Story            4   

     OverallCond  YearBuilt  YearRemodAdd RoofStyle RoofMatl Exterior1st  \
913            7       1946          1950     Gable  CompShg     MetalSd   

    Exterior2nd MasVnrType  MasVnrArea ExterQual ExterCond Foundation  \
913      CBlock       None         0.0        TA        TA      PConc   

    BsmtQual BsmtCond BsmtExposure BsmtFinType1  BsmtFinSF1 BsmtFinType2  \
913      NaN      NaN          NaN          NaN         NaN          NaN   

     BsmtFinSF2  BsmtUnfSF  TotalBsmtSF Heating HeatingQC CentralAir  \
913         NaN        NaN          NaN    GasA        TA          Y   

    Electrical  1stFlrSF  2ndFlrSF  LowQualFinSF  BsmtFullBath  BsmtHalfBath  \
913      FuseA       896         0             0           NaN           NaN   

     FullBath  HalfBath  BedroomAbvGr  KitchenAbvGr KitchenQual  TotRmsAbvGrd  \
913         1         0             2             1          TA             4   

    Functional  Fireplaces FireplaceQu GarageType  GarageYrBlt GarageFinish  \
913        Typ           0         NaN     Detchd       1946.0          Unf   

     GarageCars  GarageArea GarageQual GarageCond PavedDrive  WoodDeckSF  \
913         1.0       280.0         TA         TA          Y           0   

     OpenPorchSF  EnclosedPorch  3SsnPorch  ScreenPorch  PoolArea PoolQC  \
913            0              0          0            0         0    NaN   

     Fence MiscFeature  MiscVal  MoSold  YrSold SaleType SaleCondition  
913  MnPrv         NaN        0       4    2008    ConLD       Abnorml

Electrical


PID  GrLivArea  SalePrice  MSSubClass MSZoning  LotFrontage  \
626  916386080       1394     167500          80       RL         73.0   

     LotArea Street Alley LotShape LandContour Utilities LotConfig LandSlope  \
626     9735   Pave   NaN      Reg         Lvl    AllPub    Inside       Gtl   

    Neighborhood Condition1 Condition2 BldgType HouseStyle  OverallQual  \
626       Timber       Norm       Norm     1Fam       SLvl            5   

     OverallCond  YearBuilt  YearRemodAdd RoofStyle RoofMatl Exterior1st  \
626            5       2006          2007     Gable  CompShg     VinylSd   

    Exterior2nd MasVnrType  MasVnrArea ExterQual ExterCond Foundation  \
626     VinylSd       None         0.0        TA        TA      PConc   

    BsmtQual BsmtCond BsmtExposure BsmtFinType1  BsmtFinSF1 BsmtFinType2  \
626       Gd       TA           No          Unf         0.0          Unf   

     BsmtFinSF2  BsmtUnfSF  TotalBsmtSF Heating HeatingQC CentralAir  \
626         0.0      384.0        384.0    GasA        Gd          Y   

    Electrical  1stFlrSF  2ndFlrSF  LowQualFinSF  BsmtFullBath  BsmtHalfBath  \
626        NaN       754       640             0           0.0           0.0   

     FullBath  HalfBath  BedroomAbvGr  KitchenAbvGr KitchenQual  TotRmsAbvGrd  \
626         2         1             3             1          Gd             7   

    Functional  Fireplaces FireplaceQu GarageType  GarageYrBlt GarageFinish  \
626        Typ           0         NaN    BuiltIn       2007.0          Fin   

     GarageCars  GarageArea GarageQual GarageCond PavedDrive  WoodDeckSF  \
626         2.0       400.0         TA         TA          Y         100   

     OpenPorchSF  EnclosedPorch  3SsnPorch  ScreenPorch  PoolArea PoolQC  \
626            0              0          0            0         0    NaN   

    Fence MiscFeature  MiscVal  MoSold  YrSold SaleType SaleCondition  
626   NaN         NaN        0       5    2008      WD         Normal

BsmtFullBath


PID  GrLivArea  SalePrice  MSSubClass MSZoning  LotFrontage  \
492  908154080       3820     284700          20       RL        123.0   

     LotArea Street Alley LotShape LandContour Utilities LotConfig LandSlope  \
492    47007   Pave   NaN      IR1         Lvl    AllPub    Inside       Gtl   

    Neighborhood Condition1 Condition2 BldgType HouseStyle  OverallQual  \
492      Edwards       Norm       Norm     1Fam     1Story            5   

     OverallCond  YearBuilt  YearRemodAdd RoofStyle RoofMatl Exterior1st  \
492            7       1959          1996     Gable  CompShg     Plywood   

    Exterior2nd MasVnrType  MasVnrArea ExterQual ExterCond Foundation  \
492     Plywood       None         0.0        TA        TA       Slab   

    BsmtQual BsmtCond BsmtExposure BsmtFinType1  BsmtFinSF1 BsmtFinType2  \
492      NaN      NaN          NaN          NaN         0.0          NaN   

     BsmtFinSF2  BsmtUnfSF  TotalBsmtSF Heating HeatingQC CentralAir  \
492         0.0        0.0          0.0    GasA        TA          Y   

    Electrical  1stFlrSF  2ndFlrSF  LowQualFinSF  BsmtFullBath  BsmtHalfBath  \
492      SBrkr      3820         0             0           NaN           NaN   

     FullBath  HalfBath  BedroomAbvGr  KitchenAbvGr KitchenQual  TotRmsAbvGrd  \
492         3         1             5             1          Ex            11   

    Functional  Fireplaces FireplaceQu GarageType  GarageYrBlt GarageFinish  \
492        Typ           2          Gd     Attchd       1959.0          Unf   

     GarageCars  GarageArea GarageQual GarageCond PavedDrive  WoodDeckSF  \
492         2.0       624.0         TA         TA          Y           0   

     OpenPorchSF  EnclosedPorch  3SsnPorch  ScreenPorch  PoolArea PoolQC  \
492          372              0          0            0         0    NaN   

    Fence MiscFeature  MiscVal  MoSold  YrSold SaleType SaleCondition  
492   NaN         NaN        0       7    2008      WD         Normal

BsmtHalfBath


PID  GrLivArea  SalePrice  MSSubClass MSZoning  LotFrontage  \
492  908154080       3820     284700          20       RL        123.0   

     LotArea Street Alley LotShape LandContour Utilities LotConfig LandSlope  \
492    47007   Pave   NaN      IR1         Lvl    AllPub    Inside       Gtl   

    Neighborhood Condition1 Condition2 BldgType HouseStyle  OverallQual  \
492      Edwards       Norm       Norm     1Fam     1Story            5   

     OverallCond  YearBuilt  YearRemodAdd RoofStyle RoofMatl Exterior1st  \
492            7       1959          1996     Gable  CompShg     Plywood   

    Exterior2nd MasVnrType  MasVnrArea ExterQual ExterCond Foundation  \
492     Plywood       None         0.0        TA        TA       Slab   

    BsmtQual BsmtCond BsmtExposure BsmtFinType1  BsmtFinSF1 BsmtFinType2  \
492      NaN      NaN          NaN          NaN         0.0          NaN   

     BsmtFinSF2  BsmtUnfSF  TotalBsmtSF Heating HeatingQC CentralAir  \
492         0.0        0.0          0.0    GasA        TA          Y   

    Electrical  1stFlrSF  2ndFlrSF  LowQualFinSF  BsmtFullBath  BsmtHalfBath  \
492      SBrkr      3820         0             0           NaN           NaN   

     FullBath  HalfBath  BedroomAbvGr  KitchenAbvGr KitchenQual  TotRmsAbvGrd  \
492         3         1             5             1          Ex            11   

    Functional  Fireplaces FireplaceQu GarageType  GarageYrBlt GarageFinish  \
492        Typ           2          Gd     Attchd       1959.0          Unf   

     GarageCars  GarageArea GarageQual GarageCond PavedDrive  WoodDeckSF  \
492         2.0       624.0         TA         TA          Y           0   

     OpenPorchSF  EnclosedPorch  3SsnPorch  ScreenPorch  PoolArea PoolQC  \
492          372              0          0            0         0    NaN   

    Fence MiscFeature  MiscVal  MoSold  YrSold SaleType SaleCondition  
492   NaN         NaN        0       7    2008      WD         Normal

FireplaceQu


PID  GrLivArea  SalePrice  MSSubClass MSZoning  LotFrontage  \
456  527450210        987      83000         160       RM         21.0   

     LotArea Street Alley LotShape LandContour Utilities LotConfig LandSlope  \
456     1953   Pave   NaN      Reg         Lvl    AllPub    Inside       Gtl   

    Neighborhood Condition1 Condition2 BldgType HouseStyle  OverallQual  \
456       BrDale       Norm       Norm    Twnhs     2Story            6   

     OverallCond  YearBuilt  YearRemodAdd RoofStyle RoofMatl Exterior1st  \
456            5       1973          1973     Gable  CompShg     HdBoard   

    Exterior2nd MasVnrType  MasVnrArea ExterQual ExterCond Foundation  \
456     HdBoard    BrkFace       408.0        TA        TA     CBlock   

    BsmtQual BsmtCond BsmtExposure BsmtFinType1  BsmtFinSF1 BsmtFinType2  \
456       TA       Fa           No          BLQ       309.0          Unf   

     BsmtFinSF2  BsmtUnfSF  TotalBsmtSF Heating HeatingQC CentralAir  \
456         0.0      174.0        483.0    GasA        TA          Y   

    Electrical  1stFlrSF  2ndFlrSF  LowQualFinSF  BsmtFullBath  BsmtHalfBath  \
456      SBrkr       483       504             0           0.0           0.0   

     FullBath  HalfBath  BedroomAbvGr  KitchenAbvGr KitchenQual  TotRmsAbvGrd  \
456         1         1             2             1          TA             5   

    Functional  Fireplaces FireplaceQu GarageType  GarageYrBlt GarageFinish  \
456        Typ           0         NaN     Detchd       1973.0          Unf   

     GarageCars  GarageArea GarageQual GarageCond PavedDrive  WoodDeckSF  \
456         1.0       264.0         TA         TA          Y          72   

     OpenPorchSF  EnclosedPorch  3SsnPorch  ScreenPorch  PoolArea PoolQC  \
456            0              0          0            0         0    NaN   

    Fence MiscFeature  MiscVal  MoSold  YrSold SaleType SaleCondition  
456   NaN         NaN        0       6    2006      WD         Normal

GarageType


PID  GrLivArea  SalePrice  MSSubClass MSZoning  LotFrontage  \
316  905452070        864      99600          90       RL         50.0   

     LotArea Street Alley LotShape LandContour Utilities LotConfig LandSlope  \
316     9350   Pave   NaN      Reg         Lvl    AllPub    Inside       Gtl   

    Neighborhood Condition1 Condition2 BldgType HouseStyle  OverallQual  \
316      Edwards       Norm       Norm   Duplex     SFoyer            5   

     OverallCond  YearBuilt  YearRemodAdd RoofStyle RoofMatl Exterior1st  \
316            5       1975          1975     Gable  CompShg     HdBoard   

    Exterior2nd MasVnrType  MasVnrArea ExterQual ExterCond Foundation  \
316     HdBoard       None         0.0        TA        TA      PConc   

    BsmtQual BsmtCond BsmtExposure BsmtFinType1  BsmtFinSF1 BsmtFinType2  \
316       Gd       TA           Av          GLQ       864.0          Unf   

     BsmtFinSF2  BsmtUnfSF  TotalBsmtSF Heating HeatingQC CentralAir  \
316         0.0        0.0        864.0    GasA        Fa          N   

    Electrical  1stFlrSF  2ndFlrSF  LowQualFinSF  BsmtFullBath  BsmtHalfBath  \
316      SBrkr       864         0             0           1.0           0.0   

     FullBath  HalfBath  BedroomAbvGr  KitchenAbvGr KitchenQual  TotRmsAbvGrd  \
316         1         0             2             1          TA             4   

    Functional  Fireplaces FireplaceQu GarageType  GarageYrBlt GarageFinish  \
316        Typ           0         NaN        NaN          NaN          NaN   

     GarageCars  GarageArea GarageQual GarageCond PavedDrive  WoodDeckSF  \
316         0.0         0.0        NaN        NaN          Y           0   

     OpenPorchSF  EnclosedPorch  3SsnPorch  ScreenPorch  PoolArea PoolQC  \
316            0              0          0            0         0    NaN   

    Fence MiscFeature  MiscVal  MoSold  YrSold SaleType SaleCondition  
316   NaN         NaN        0       4    2008      WD         Normal

GarageYrBlt


PID  GrLivArea  SalePrice  MSSubClass MSZoning  LotFrontage  \
373  902128075       1027     120000          30       RM         60.0   

     LotArea Street Alley LotShape LandContour Utilities LotConfig LandSlope  \
373     5400   Pave   NaN      Reg         Lvl    AllPub    Inside       Gtl   

    Neighborhood Condition1 Condition2 BldgType HouseStyle  OverallQual  \
373      OldTown       Norm       Norm     1Fam     1Story            7   

     OverallCond  YearBuilt  YearRemodAdd RoofStyle RoofMatl Exterior1st  \
373            6       1920          2006     Gable  CompShg      Stucco   

    Exterior2nd MasVnrType  MasVnrArea ExterQual ExterCond Foundation  \
373      Stucco       None         0.0        Gd        TA     CBlock   

    BsmtQual BsmtCond BsmtExposure BsmtFinType1  BsmtFinSF1 BsmtFinType2  \
373       TA       TA           No          Unf         0.0          Unf   

     BsmtFinSF2  BsmtUnfSF  TotalBsmtSF Heating HeatingQC CentralAir  \
373         0.0      931.0        931.0    GasA        TA          Y   

    Electrical  1stFlrSF  2ndFlrSF  LowQualFinSF  BsmtFullBath  BsmtHalfBath  \
373      SBrkr      1027         0             0           0.0           1.0   

     FullBath  HalfBath  BedroomAbvGr  KitchenAbvGr KitchenQual  TotRmsAbvGrd  \
373         1         0             2             1          Gd             5   

    Functional  Fireplaces FireplaceQu GarageType  GarageYrBlt GarageFinish  \
373        Typ           1          TA        NaN          NaN          NaN   

     GarageCars  GarageArea GarageQual GarageCond PavedDrive  WoodDeckSF  \
373         0.0         0.0        NaN        NaN          N           0   

     OpenPorchSF  EnclosedPorch  3SsnPorch  ScreenPorch  PoolArea PoolQC  \
373           28              0          0            0         0    NaN   

    Fence MiscFeature  MiscVal  MoSold  YrSold SaleType SaleCondition  
373   NaN         NaN        0       2    2008      WD         Normal

GarageFinish


PID  GrLivArea  SalePrice  MSSubClass MSZoning  LotFrontage  \
25  902104060       1140      96500          50       RM         55.0   

    LotArea Street Alley LotShape LandContour Utilities LotConfig LandSlope  \
25     8800   Pave  Grvl      Reg         Lvl    AllPub    Corner       Gtl   

   Neighborhood Condition1 Condition2 BldgType HouseStyle  OverallQual  \
25      OldTown       Norm       Norm     1Fam     1.5Fin            4   

    OverallCond  YearBuilt  YearRemodAdd RoofStyle RoofMatl Exterior1st  \
25            7       1910          2004     Gable  CompShg     Wd Sdng   

   Exterior2nd MasVnrType  MasVnrArea ExterQual ExterCond Foundation BsmtQual  \
25     Wd Sdng       None         0.0        TA        TA     BrkTil       TA   

   BsmtCond BsmtExposure BsmtFinType1  BsmtFinSF1 BsmtFinType2  BsmtFinSF2  \
25       Fa           No          Unf         0.0          Unf         0.0   

    BsmtUnfSF  TotalBsmtSF Heating HeatingQC CentralAir Electrical  1stFlrSF  \
25      576.0        576.0    GasA        Gd          Y      SBrkr       792   

    2ndFlrSF  LowQualFinSF  BsmtFullBath  BsmtHalfBath  FullBath  HalfBath  \
25       348             0           0.0           0.0         1         0   

    BedroomAbvGr  KitchenAbvGr KitchenQual  TotRmsAbvGrd Functional  \
25             3             1          TA             7       Min2   

    Fireplaces FireplaceQu GarageType  GarageYrBlt GarageFinish  GarageCars  \
25           0         NaN        NaN          NaN          NaN         0.0   

    GarageArea GarageQual GarageCond PavedDrive  WoodDeckSF  OpenPorchSF  \
25         0.0        NaN        NaN          N           0          160   

    EnclosedPorch  3SsnPorch  ScreenPorch  PoolArea PoolQC Fence MiscFeature  \
25              0          0            0         0    NaN   NaN         NaN   

    MiscVal  MoSold  YrSold SaleType SaleCondition  
25        0       6    2010      WD         Normal

GarageCars


PID  GrLivArea  SalePrice  MSSubClass MSZoning  LotFrontage  \
434  910201180       1828     150909          70       RM         50.0   

     LotArea Street Alley LotShape LandContour Utilities LotConfig LandSlope  \
434     9060   Pave   NaN      Reg         Lvl    AllPub    Inside       Gtl   

    Neighborhood Condition1 Condition2 BldgType HouseStyle  OverallQual  \
434       IDOTRR       Norm       Norm     1Fam     2Story            5   

     OverallCond  YearBuilt  YearRemodAdd RoofStyle RoofMatl Exterior1st  \
434            6       1923          1999     Gable  CompShg     Wd Sdng   

    Exterior2nd MasVnrType  MasVnrArea ExterQual ExterCond Foundation  \
434     Plywood       None         0.0        TA        TA     BrkTil   

    BsmtQual BsmtCond BsmtExposure BsmtFinType1  BsmtFinSF1 BsmtFinType2  \
434       Gd       TA           No          ALQ       548.0          Unf   

     BsmtFinSF2  BsmtUnfSF  TotalBsmtSF Heating HeatingQC CentralAir  \
434         0.0      311.0        859.0    GasA        Ex          Y   

    Electrical  1stFlrSF  2ndFlrSF  LowQualFinSF  BsmtFullBath  BsmtHalfBath  \
434      SBrkr       942       886             0           0.0           0.0   

     FullBath  HalfBath  BedroomAbvGr  KitchenAbvGr KitchenQual  TotRmsAbvGrd  \
434         2         0             3             1          Gd             6   

    Functional  Fireplaces FireplaceQu GarageType  GarageYrBlt GarageFinish  \
434        Typ           0         NaN     Detchd          NaN          NaN   

     GarageCars  GarageArea GarageQual GarageCond PavedDrive  WoodDeckSF  \
434         NaN         NaN        NaN        NaN          Y         174   

     OpenPorchSF  EnclosedPorch  3SsnPorch  ScreenPorch  PoolArea PoolQC  \
434            0            212          0            0         0    NaN   

     Fence MiscFeature  MiscVal  MoSold  YrSold SaleType SaleCondition  
434  MnPrv         NaN        0       3    2007      WD         Alloca

GarageArea


PID  GrLivArea  SalePrice  MSSubClass MSZoning  LotFrontage  \
434  910201180       1828     150909          70       RM         50.0   

     LotArea Street Alley LotShape LandContour Utilities LotConfig LandSlope  \
434     9060   Pave   NaN      Reg         Lvl    AllPub    Inside       Gtl   

    Neighborhood Condition1 Condition2 BldgType HouseStyle  OverallQual  \
434       IDOTRR       Norm       Norm     1Fam     2Story            5   

     OverallCond  YearBuilt  YearRemodAdd RoofStyle RoofMatl Exterior1st  \
434            6       1923          1999     Gable  CompShg     Wd Sdng   

    Exterior2nd MasVnrType  MasVnrArea ExterQual ExterCond Foundation  \
434     Plywood       None         0.0        TA        TA     BrkTil   

    BsmtQual BsmtCond BsmtExposure BsmtFinType1  BsmtFinSF1 BsmtFinType2  \
434       Gd       TA           No          ALQ       548.0          Unf   

     BsmtFinSF2  BsmtUnfSF  TotalBsmtSF Heating HeatingQC CentralAir  \
434         0.0      311.0        859.0    GasA        Ex          Y   

    Electrical  1stFlrSF  2ndFlrSF  LowQualFinSF  BsmtFullBath  BsmtHalfBath  \
434      SBrkr       942       886             0           0.0           0.0   

     FullBath  HalfBath  BedroomAbvGr  KitchenAbvGr KitchenQual  TotRmsAbvGrd  \
434         2         0             3             1          Gd             6   

    Functional  Fireplaces FireplaceQu GarageType  GarageYrBlt GarageFinish  \
434        Typ           0         NaN     Detchd          NaN          NaN   

     GarageCars  GarageArea GarageQual GarageCond PavedDrive  WoodDeckSF  \
434         NaN         NaN        NaN        NaN          Y         174   

     OpenPorchSF  EnclosedPorch  3SsnPorch  ScreenPorch  PoolArea PoolQC  \
434            0            212          0            0         0    NaN   

     Fence MiscFeature  MiscVal  MoSold  YrSold SaleType SaleCondition  
434  MnPrv         NaN        0       3    2007      WD         Alloca

GarageQual


PID  GrLivArea  SalePrice  MSSubClass MSZoning  LotFrontage  \
769  903454020        819      72000          30       RM         50.0   

     LotArea Street Alley LotShape LandContour Utilities LotConfig LandSlope  \
769     9000   Pave   NaN      Reg         Lvl    AllPub    Inside       Gtl   

    Neighborhood Condition1 Condition2 BldgType HouseStyle  OverallQual  \
769       IDOTRR       Norm       Norm     1Fam     1Story            5   

     OverallCond  YearBuilt  YearRemodAdd RoofStyle RoofMatl Exterior1st  \
769            4       1919          1950     Gable  CompShg     Wd Sdng   

    Exterior2nd MasVnrType  MasVnrArea ExterQual ExterCond Foundation  \
769     Wd Sdng       None         0.0        TA        TA     BrkTil   

    BsmtQual BsmtCond BsmtExposure BsmtFinType1  BsmtFinSF1 BsmtFinType2  \
769       TA       TA           No          Unf         0.0          Unf   

     BsmtFinSF2  BsmtUnfSF  TotalBsmtSF Heating HeatingQC CentralAir  \
769         0.0      610.0        610.0    GasA        Ex          N   

    Electrical  1stFlrSF  2ndFlrSF  LowQualFinSF  BsmtFullBath  BsmtHalfBath  \
769      FuseA       819         0             0           0.0           0.0   

     FullBath  HalfBath  BedroomAbvGr  KitchenAbvGr KitchenQual  TotRmsAbvGrd  \
769         1         0             2             1          Gd             4   

    Functional  Fireplaces FireplaceQu GarageType  GarageYrBlt GarageFinish  \
769        Typ           0         NaN        NaN          NaN          NaN   

     GarageCars  GarageArea GarageQual GarageCond PavedDrive  WoodDeckSF  \
769         0.0         0.0        NaN        NaN          N           0   

     OpenPorchSF  EnclosedPorch  3SsnPorch  ScreenPorch  PoolArea PoolQC  \
769            0              0          0            0         0    NaN   

    Fence MiscFeature  MiscVal  MoSold  YrSold SaleType SaleCondition  
769   NaN         NaN        0       9    2006      WD        Abnorml

GarageCond


PID  GrLivArea  SalePrice  MSSubClass MSZoning  LotFrontage  \
483  905101300       1832     136000          90       RL         72.0   

     LotArea Street Alley LotShape LandContour Utilities LotConfig LandSlope  \
483    10773   Pave   NaN      Reg         Lvl    AllPub    Inside       Gtl   

    Neighborhood Condition1 Condition2 BldgType HouseStyle  OverallQual  \
483       Sawyer       Norm       Norm   Duplex     1Story            4   

     OverallCond  YearBuilt  YearRemodAdd RoofStyle RoofMatl Exterior1st  \
483            3       1967          1967     Gable  Tar&Grv     Plywood   

    Exterior2nd MasVnrType  MasVnrArea ExterQual ExterCond Foundation  \
483     Plywood    BrkFace        72.0        Fa        Fa     CBlock   

    BsmtQual BsmtCond BsmtExposure BsmtFinType1  BsmtFinSF1 BsmtFinType2  \
483       TA       TA           No          ALQ       704.0          Unf   

     BsmtFinSF2  BsmtUnfSF  TotalBsmtSF Heating HeatingQC CentralAir  \
483         0.0     1128.0       1832.0    GasA        TA          N   

    Electrical  1stFlrSF  2ndFlrSF  LowQualFinSF  BsmtFullBath  BsmtHalfBath  \
483      SBrkr      1832         0             0           2.0           0.0   

     FullBath  HalfBath  BedroomAbvGr  KitchenAbvGr KitchenQual  TotRmsAbvGrd  \
483         2         0             4             2          TA             8   

    Functional  Fireplaces FireplaceQu GarageType  GarageYrBlt GarageFinish  \
483        Typ           0         NaN        NaN          NaN          NaN   

     GarageCars  GarageArea GarageQual GarageCond PavedDrive  WoodDeckSF  \
483         0.0         0.0        NaN        NaN          Y           0   

     OpenPorchSF  EnclosedPorch  3SsnPorch  ScreenPorch  PoolArea PoolQC  \
483           58              0          0            0         0    NaN   

    Fence MiscFeature  MiscVal  MoSold  YrSold SaleType SaleCondition  
483   NaN         NaN        0       5    2010      WD         Normal

PoolQC


PID  GrLivArea  SalePrice  MSSubClass MSZoning  LotFrontage  \
767  902405120       1480     119000          50       RM         60.0   

     LotArea Street Alley LotShape LandContour Utilities LotConfig LandSlope  \
767     5400   Pave   NaN      Reg         Lvl    AllPub    Inside       Gtl   

    Neighborhood Condition1 Condition2 BldgType HouseStyle  OverallQual  \
767      OldTown       Norm       Norm     1Fam     1.5Fin            6   

     OverallCond  YearBuilt  YearRemodAdd RoofStyle RoofMatl Exterior1st  \
767            6       1920          1950     Gable  CompShg     MetalSd   

    Exterior2nd MasVnrType  MasVnrArea ExterQual ExterCond Foundation  \
767     MetalSd       None         0.0        Fa        TA     BrkTil   

    BsmtQual BsmtCond BsmtExposure BsmtFinType1  BsmtFinSF1 BsmtFinType2  \
767       TA       TA           No          Unf         0.0          Unf   

     BsmtFinSF2  BsmtUnfSF  TotalBsmtSF Heating HeatingQC CentralAir  \
767         0.0      972.0        972.0    GasA        Ex          Y   

    Electrical  1stFlrSF  2ndFlrSF  LowQualFinSF  BsmtFullBath  BsmtHalfBath  \
767      SBrkr      1044         0           436           0.0           0.0   

     FullBath  HalfBath  BedroomAbvGr  KitchenAbvGr KitchenQual  TotRmsAbvGrd  \
767         1         0             3             1          TA             6   

    Functional  Fireplaces FireplaceQu GarageType  GarageYrBlt GarageFinish  \
767        Typ           0         NaN     Detchd       1920.0          Unf   

     GarageCars  GarageArea GarageQual GarageCond PavedDrive  WoodDeckSF  \
767         1.0       207.0         Fa         TA          Y           0   

     OpenPorchSF  EnclosedPorch  3SsnPorch  ScreenPorch  PoolArea PoolQC  \
767            0            176          0            0         0    NaN   

    Fence MiscFeature  MiscVal  MoSold  YrSold SaleType SaleCondition  
767   NaN         NaN        0       9    2009    ConLI        Family

Fence


PID  GrLivArea  SalePrice  MSSubClass MSZoning  LotFrontage  \
480  924100050       1479     161000          50       RL         68.0   

     LotArea Street Alley LotShape LandContour Utilities LotConfig LandSlope  \
480    10544   Pave   NaN      IR1         Lvl    AllPub    Inside       Mod   

    Neighborhood Condition1 Condition2 BldgType HouseStyle  OverallQual  \
480      Mitchel       Norm       Norm     1Fam     1.5Fin            5   

     OverallCond  YearBuilt  YearRemodAdd RoofStyle RoofMatl Exterior1st  \
480            5       1969          1969     Gable  CompShg     HdBoard   

    Exterior2nd MasVnrType  MasVnrArea ExterQual ExterCond Foundation  \
480     HdBoard       None         0.0        TA        TA     CBlock   

    BsmtQual BsmtCond BsmtExposure BsmtFinType1  BsmtFinSF1 BsmtFinType2  \
480       TA       TA           Av          BLQ       476.0          Unf   

     BsmtFinSF2  BsmtUnfSF  TotalBsmtSF Heating HeatingQC CentralAir  \
480         0.0      388.0        864.0    GasA        TA          Y   

    Electrical  1stFlrSF  2ndFlrSF  LowQualFinSF  BsmtFullBath  BsmtHalfBath  \
480      SBrkr       864       615             0           0.0           0.0   

     FullBath  HalfBath  BedroomAbvGr  KitchenAbvGr KitchenQual  TotRmsAbvGrd  \
480         2         0             5             1          TA             8   

    Functional  Fireplaces FireplaceQu GarageType  GarageYrBlt GarageFinish  \
480        Typ           0         NaN     Attchd       1969.0          Fin   

     GarageCars  GarageArea GarageQual GarageCond PavedDrive  WoodDeckSF  \
480         1.0       275.0         TA         TA          Y         287   

     OpenPorchSF  EnclosedPorch  3SsnPorch  ScreenPorch  PoolArea PoolQC  \
480            0            280          0            0         0    NaN   

    Fence MiscFeature  MiscVal  MoSold  YrSold SaleType SaleCondition  
480   NaN         NaN        0       4    2010      WD         Normal

MiscFeature


PID  GrLivArea  SalePrice  MSSubClass MSZoning  LotFrontage  \
377  528250100       1430     180000          80       RL          NaN   

     LotArea Street Alley LotShape LandContour Utilities LotConfig LandSlope  \
377     7750   Pave   NaN      IR1         Lvl    AllPub    Inside       Gtl   

    Neighborhood Condition1 Condition2 BldgType HouseStyle  OverallQual  \
377      Gilbert       Norm       Norm     1Fam       SLvl            7   

     OverallCond  YearBuilt  YearRemodAdd RoofStyle RoofMatl Exterior1st  \
377            5       2000          2000     Gable  CompShg     VinylSd   

    Exterior2nd MasVnrType  MasVnrArea ExterQual ExterCond Foundation  \
377     VinylSd       None         0.0        TA        TA      PConc   

    BsmtQual BsmtCond BsmtExposure BsmtFinType1  BsmtFinSF1 BsmtFinType2  \
377       Gd       TA           No          GLQ       250.0          Unf   

     BsmtFinSF2  BsmtUnfSF  TotalBsmtSF Heating HeatingQC CentralAir  \
377         0.0      134.0        384.0    GasA        Ex          Y   

    Electrical  1stFlrSF  2ndFlrSF  LowQualFinSF  BsmtFullBath  BsmtHalfBath  \
377      SBrkr       774       656             0           0.0           0.0   

     FullBath  HalfBath  BedroomAbvGr  KitchenAbvGr KitchenQual  TotRmsAbvGrd  \
377         2         1             3             1          TA             7   

    Functional  Fireplaces FireplaceQu GarageType  GarageYrBlt GarageFinish  \
377        Typ           1          TA    BuiltIn       2000.0          Fin   

     GarageCars  GarageArea GarageQual GarageCond PavedDrive  WoodDeckSF  \
377         2.0       400.0         TA         TA          Y         180   

     OpenPorchSF  EnclosedPorch  3SsnPorch  ScreenPorch  PoolArea PoolQC  \
377            0              0          0            0         0    NaN   

    Fence MiscFeature  MiscVal  MoSold  YrSold SaleType SaleCondition  
377   NaN         NaN        0       4    2010      WD         Normal

Notes: 
- some houses without basements record basement square footage as NaN, others record basement square footage as 0. When adding basement and above-grade data, make sure to not yield misleading null values
- most missing data appear to be due to inapplicability, but there are some discrepancies such as 127 missing garage type features and 129 missing garage year built and garage finish features. 

### Basic feature engineering

In [151]:
housedf['ExteriorSF'] = (housedf['LotArea'] - housedf['1stFlrSF'])
housedf['AgeAtSale'] = (housedf['YrSold'] - housedf['YearBuilt'])
housedf['TotalFullBath'] = (housedf['FullBath'] + housedf['BsmtFullBath'].fillna(0))
housedf['TotalHalfBath'] = (housedf['HalfBath'] + housedf['BsmtHalfBath'].fillna(0))

### Joining Parcel Data

In [152]:
parcel_data_raw = pd.read_csv('./data/Ames Real Estate Data.csv', index_col=0)

/var/folders/j2/lf9zt3017c3fz8rrnty2f8h80000gn/T/ipykernel_72021/1795626433.py:1: DtypeWarning: Columns (36,40,81,87) have mixed types. Specify dtype option on import or set low_memory=False.
  parcel_data_raw = pd.read_csv('./data/Ames Real Estate Data.csv', index_col=0)


In [153]:
# error indicates that the following columns have mixed data types
parcel_data_raw.columns[[36, 40, 81, 87]]

Index(['X1TPr_S', 'LndAcX1S', 'PA-PreD', 'Date'], dtype='object')

In [154]:
parcel_data_raw.head(3)

GeoRefNo  Tier  Range                 Prop_Addr ZngCdPr ZngCdSc  \
MapRefNo                                                                      
520400001  520400001     0      0  4599 GEORGE W CARVER AVE   FS-RL     NaN   
520440010  520440010     0      0         4507 EVEREST  AVE   FS-RL     NaN   
520440020  520440020     0      0         4513 EVEREST  AVE   FS-RL     NaN   

          ZngOLPr ZngOLSc    ClassPr_S                             ClassSc_S  \
MapRefNo                                                                       
520400001     NaN     NaN  RESIDENTIAL  VACANT LAND - IMPROVED OR UNIMPROVED   
520440010     NaN     NaN  RESIDENTIAL  VACANT LAND - IMPROVED OR UNIMPROVED   
520440020     NaN     NaN  RESIDENTIAL  VACANT LAND - IMPROVED OR UNIMPROVED   

                                             Legal_Pr  SchD_S  TxD_S  \
MapRefNo                                                               
520400001  SCENIC VALLEY SD 4TH ADD \nOUTLOT V\n AMES       5     45   
520440010     SCENIC VALLEY SD 4TH ADD LOT:16 \n AMES       5     45   
520440020     SCENIC VALLEY SD 4TH ADD LOT:17 \n AMES       5     45   

                                 MA_Ownr1 MA_Ownr2  \
MapRefNo                                             
520400001  HUNZIKER DEVELOPMENT GROUP LLC      NaN   
520440010  HUNZIKER DEVELOPMENT GROUP LLC      NaN   
520440020                 HAPPE HOMES LLP      NaN   

                             MA_Line1  MA_Line2 MA_City MA_State  MA_Zip1  \
MapRefNo                                                                    
520400001               105 S 16TH ST       NaN    AMES       IA  50010.0   
520440010               105 S 16TH ST       NaN    AMES       IA  50010.0   
520440020  2575 N ANKENY BLVD STE 211       NaN  ANKENY       IA  50023.0   

           MA_Zip2  Rcrd_Yr  Rcrd_Mo    Inst1_No  Inst1_Yr  Inst1_Mo Inst1TPr  \
MapRefNo                                                                        
520400001      NaN   2019.0      7.0   201906213    2019.0       7.0      NaN   
520440010      NaN   2019.0      7.0   201906213    2019.0       7.0      NaN   
520440020      NaN   2020.0      5.0  2020-04989    2020.0       5.0      WRD   

           LndAc_S  ImpAc_S  OthAc_S  TtlVal_AsrYr ValType X1TPr_D X1TSc_D  \
MapRefNo                                                                     
520400001   481100        0        0        481100       L     NaN     NaN   
520440010     5700        0        0          5700       L     NaN     NaN   
520440020     4400        0        0          4400       L     NaN     NaN   

           X2TPr_D X2TSc_D X1TPr_S X1TSc_S  X2TPr_S X2TSc_S  LndAcX1S  \
MapRefNo                                                                
520400001      NaN     NaN     NaN     NaN      NaN     NaN         0   
520440010      NaN     NaN     NaN     NaN      NaN     NaN         0   
520440020      NaN     NaN     NaN     NaN      NaN     NaN         0   

           ImpAcX1S  ImpAcX2S  HSTtl_D  MilVal_D  HSTtl_S  MilVal_S  AcreX_S1  \
MapRefNo                                                                        
520400001         0         0      NaN       NaN      NaN       NaN       0.0   
520440010         0         0      NaN       NaN      NaN       NaN       0.0   
520440020         0         0      NaN       NaN      NaN       NaN       0.0   

           AcreGr  AcreNt_S Neighborhood  LotArea ParType  BldgNo_S  DwlgNo_S  \
MapRefNo                                                                        
520400001     0.0       0.0          NaN      NaN     NaN         0         0   
520440010     0.0       0.0          NaN      NaN     NaN         0         0   
520440020     0.0       0.0          NaN      NaN     NaN         0         0   

          BldgType  YrBuilt HouseStyle Foundation RoofMatl Ext1 Ext2  \
MapRefNo                                                               
520400001      NaN      NaN        NaN        NaN      NaN  NaN  NaN   
520440

In [155]:
parcel_data_raw.shape

(22213, 90)

In [156]:
# null value counts
pd.DataFrame(pd.isnull(parcel_data_raw).sum(axis=0)).T

GeoRefNo  Tier  Range  Prop_Addr  ZngCdPr  ZngCdSc  ZngOLPr  ZngOLSc  \
0         0     0      0         20      106    22187    20243    22060   

   ClassPr_S  ClassSc_S  Legal_Pr  SchD_S  TxD_S  MA_Ownr1  MA_Ownr2  \
0          0          0         0       0      0         0     19337   

   MA_Line1  MA_Line2  MA_City  MA_State  MA_Zip1  MA_Zip2  Rcrd_Yr  Rcrd_Mo  \
0        14     22213       14        14       14     8890     1907     1907   

   Inst1_No  Inst1_Yr  Inst1_Mo  Inst1TPr  LndAc_S  ImpAc_S  OthAc_S  \
0      1857      1907      1907      2677        0        0        0   

   TtlVal_AsrYr  ValType  X1TPr_D  X1TSc_D  X2TPr_D  X2TSc_D  X1TPr_S  \
0             0        0    21519    21519    22212    22212    21522   

   X1TSc_S  X2TPr_S  X2TSc_S  LndAcX1S  ImpAcX1S  ImpAcX2S  HSTtl_D  MilVal_D  \
0    21522    22212    22212         0         0         0    13024     13024   

   HSTtl_S  MilVal_S  AcreX_S1  AcreGr  AcreNt_S  Neighborhood  LotArea  \
0    12966     12966         0       0         0          4308     4308   

   ParType  BldgNo_S  DwlgNo_S  BldgType  YrBuilt  HouseStyle  Foundation  \
0     4308         0         0      4308     4308        4308        4308   

   RoofMatl  Ext1  Ext2  MasVnrType  Heating  Central Air   GLA  TtlBsmtSF  \
0      4308  4308  4308        4308     4308         4308  4308       4308   

   TotRmsAbvGrd  Fireplaces  PoolArea  GarageType  GarYrBlt  Cars  GarageArea  \
0          4308        4308      4308        4308      4308  4308        4308   

   YrSold_YYYY  MoSold_MM  SalePrice  SaleType  SaleCond  ParclRel  PA-Nmbr  \
0        19247      19247      19247     19247     19247         0       20   

   PA-PreD  PA-Strt  PA-StSfx  PA-PostD  PA-UnTyp  PA-UntNo  Date  Source  \
0    20437       20        98     22213     14468     14096     0       0   

   NmbrBRs  
0     4308

We'll take just the address and the owner names

In [157]:
parcel_data = parcel_data_raw[['Prop_Addr', 'MA_Ownr1', 'MA_Ownr2']]

In [158]:
housedf = pd.merge(housedf, parcel_data, left_on='PID', right_index=True)
housedf = housedf.reset_index()
housedf.head(5)
# Note: at this point, housedf's lenght becomes 2603?

index        PID  GrLivArea  SalePrice  MSSubClass MSZoning  LotFrontage  \
0      1  909176150        856     126000          30       RL          NaN   
1      2  905476230       1049     139500         120       RL         42.0   
2      2  905476230       1049     139500         120       RL         42.0   
3      3  911128020       1001     124900          30  C (all)         60.0   
4      4  535377150       1039     114000          70       RL         80.0   

   LotArea Street Alley LotShape LandContour Utilities LotConfig LandSlope  \
0     7890   Pave   NaN      Reg         Lvl    AllPub    Corner       Gtl   
1     4235   Pave   NaN      Reg         Lvl    AllPub    Inside       Gtl   
2     4235   Pave   NaN      Reg         Lvl    AllPub    Inside       Gtl   
3     6060   Pave   NaN      Reg         Lvl    AllPub    Inside       Gtl   
4     8146   Pave   NaN      Reg         Lvl    AllPub    Corner       Gtl   

  Neighborhood Condition1 Condition2 BldgType HouseStyle  OverallQual  \
0        SWISU       Norm       Norm     1Fam     1Story            6   
1      Edwards       Norm       Norm   TwnhsE     1Story            5   
2      Edwards       Norm       Norm   TwnhsE     1Story            5   
3       IDOTRR       Norm       Norm     1Fam     1Story            5   
4      OldTown       Norm       Norm     1Fam     2Story            4   

   OverallCond  YearBuilt  YearRemodAdd RoofStyle RoofMatl Exterior1st  \
0            6       1939          1950     Gable  CompShg     Wd Sdng   
1            5       1984          1984     Gable  CompShg     HdBoard   
2            5       1984          1984     Gable  CompShg     HdBoard   
3            9       1930          2007       Hip  CompShg     MetalSd   
4            8       1900          2003     Gable  CompShg     MetalSd   

  Exterior2nd MasVnrType  MasVnrArea ExterQual ExterCond Foundation BsmtQual  \
0     Wd Sdng       None         0.0        TA        TA     CBlock       TA   
1     HdBoard    BrkFace       149.0        Gd        TA     CBlock       Gd   
2     HdBoard    BrkFace       149.0        Gd        TA     CBlock       Gd   
3     MetalSd       None         0.0        Gd        TA     BrkTil       TA   
4     MetalSd       None         0.0        Gd        Gd     BrkTil       Fa   

  BsmtCond BsmtExposure BsmtFinType1  BsmtFinSF1 BsmtFinType2  BsmtFinSF2  \
0       TA           No          Rec       238.0          Unf         0.0   
1       TA           Mn          GLQ       552.0          ALQ       393.0   
2       TA           Mn          GLQ       552.0          ALQ       393.0   
3       TA           No          ALQ       737.0          Unf         0.0   
4       TA           No          Unf         0.0          Unf         0.0   

   BsmtUnfSF  TotalBsmtSF Heating HeatingQC CentralAir Electrical  1stFlrSF  \
0      618.0        856.0    GasA        TA          Y      SBrkr       856   
1      104.0       1049.0    GasA        TA          Y      SBrkr      1049   
2      104.0       1049.0    GasA        TA          Y      SBrkr      1049   
3      100.0        837.0    GasA        Ex          Y      SBrkr      1001   
4      405.0        405.0    GasA        Gd          Y      SBrkr       717   

   2ndFlrSF  LowQualFinSF  BsmtFullBath  BsmtHalfBath  FullBath  HalfBath  \
0         0             0           1.0           0.0         1         0   
1         0             0           1.0           0.0         2         0   
2         0             0           1.0           0.0         2         0   
3         0             0           0.0           0.0         1         0   
4       322             0           0.0           0.0         1         0   

   BedroomAbvGr  KitchenAbvGr KitchenQual  TotRmsAbvGrd Functional  \
0             2             1          TA             4        Typ   
1             2             1          Gd             5        Typ   
2             2             1          Gd             5        Typ   
3            

In [108]:
import geopy
from geopy import Nominatim
from geopy.extra.rate_limiter import RateLimiter

locator = Nominatim(user_agent="myGeocoder")
geocode = RateLimiter(locator.geocode, min_delay_seconds=1.5)

In [109]:
# note: this takes a long time
locations = [geocode(addr+', Ames, USA') if type(addr)==str else None for addr in housedf['Prop_Addr']]

In [116]:
locations_df = pd.DataFrame({'Address': [location.address if location is not None else None for location in locations],
                             'Latitude': [location.latitude if location is not None else None for location in locations],
                             'Longitude': [location.longitude if location is not None else None for location in locations]})

In [159]:
housedf['Received Address'] = locations_df['Address']
housedf['Latitude'] = locations_df['Latitude']
housedf['Longitude'] = locations_df['Longitude']

In [160]:
# sanity check that addresses match up
housedf[['Prop_Addr', 'Received Address', 'Latitude', 'Longitude']]

Prop_Addr                                   Received Address  \
0         436 HAYWARD AVE  436, Hayward Avenue, Union Drive, Ames, Story ...   
1            3416 WEST ST  3416, West Street, Union Drive, Ames, Story Co...   
2            3416 WEST ST  3416, West Street, Union Drive, Ames, Story Co...   
3            320 S 2ND ST  320, South 2nd Street, Ames, Story County, Iow...   
4        1524 DOUGLAS AVE  1524, Douglas Avenue, Edgewood Park, Ames, Sto...   
...                   ...                                                ...   
2598   1021 RIDGEWOOD AVE  1021, Ridgewood Avenue, Ridgewood, Ames, Story...   
2599        3619 MARY CIR  3619, Mary Circle, Ames, Story County, Iowa, 5...   
2600  2140 SUNSET DR 2142  2140, Sunset Drive, Greekland, Ames, Story Cou...   
2601    5319 CLEMENS BLVD  5319, Clemens Boulevard, Bentwood, Ames, Story...   
2602      1025 VERMONT CT  1025, Vermont Court, Valley East, Ontario, Ame...   

       Latitude  Longitude  
0     42.018564 -93.651619  
1     42.024855 -93.663671  
2     42.024855 -93.663671  
3     42.021548 -93.614068  
4     42.037391 -93.612207  
...         ...        ...  
2598  42.032102 -93.626373  
2599  42.027669 -93.666138  
2600  42.020129 -93.643832  
2601  42.016802 -93.689748  
2602  42.030997 -93.682220  

[2603 rows x 4 columns]

In [165]:
housedf = housedf.drop(['Received Address', 'index'], axis=1)

In [166]:
housedf

PID  GrLivArea  SalePrice  MSSubClass MSZoning  LotFrontage  \
0     909176150        856     126000          30       RL          NaN   
1     905476230       1049     139500         120       RL         42.0   
2     905476230       1049     139500         120       RL         42.0   
3     911128020       1001     124900          30  C (all)         60.0   
4     535377150       1039     114000          70       RL         80.0   
...         ...        ...        ...         ...      ...          ...   
2598  903205040        952     121000          30       RL          NaN   
2599  905402060       1733     139600          20       RL          NaN   
2600  909275030       2002     145000          90       RH         82.0   
2601  907192040       1842     217500          60       RL          NaN   
2602  906223180       1911     215000          60       RL         80.0   

      LotArea Street Alley LotShape LandContour Utilities LotConfig LandSlope  \
0        7890   Pave   NaN      Reg         Lvl    AllPub    Corner       Gtl   
1        4235   Pave   NaN      Reg         Lvl    AllPub    Inside       Gtl   
2        4235   Pave   NaN      Reg         Lvl    AllPub    Inside       Gtl   
3        6060   Pave   NaN      Reg         Lvl    AllPub    Inside       Gtl   
4        8146   Pave   NaN      Reg         Lvl    AllPub    Corner       Gtl   
...       ...    ...   ...      ...         ...       ...       ...       ...   
2598     8854   Pave   NaN      Reg         Lvl    AllPub    Inside       Gtl   
2599    13680   Pave   NaN      IR1         Lvl    AllPub   CulDSac       Gtl   
2600     6270   Pave   NaN      Reg         HLS    AllPub    Inside       Gtl   
2601     8826   Pave   NaN      Reg         Lvl    AllPub    Inside       Gtl   
2602     9554   Pave   NaN      IR1         Lvl    AllPub    Inside       Gtl   

     Neighborhood Condition1 Condition2 BldgType HouseStyle  OverallQual  \
0           SWISU       Norm       Norm     1Fam     1Story            6   
1         Edwards       Norm       Norm   TwnhsE     1Story            5   
2         Edwards       Norm       Norm   TwnhsE     1Story            5   
3          IDOTRR       Norm       Norm     1Fam     1Story            5   
4         OldTown       Norm       Norm     1Fam     2Story            4   
...           ...        ...        ...      ...        ...          ...   
2598      BrkSide       Norm       Norm     1Fam     1.5Unf            6   
2599      Edwards       Norm       Norm     1Fam     1Story            3   
2600      Crawfor       Norm       Norm   Duplex     2Story            5   
2601      CollgCr       Norm       Norm     1Fam     2Story            7   
2602      SawyerW       Norm       Norm     1Fam     2Story            8   

      OverallCond  YearBuilt  YearRemodAdd RoofStyle RoofMatl Exterior1st  \
0               6       1939          1950     Gable  CompShg     Wd Sdng   
1               5       1984          1984     Gable  CompShg     HdBoard   
2               5       1984          1984     Gable  CompShg     HdBoard   
3               9       1930          2007       Hip  CompShg     MetalSd   
4               8       1900          2003     Gable  CompShg     MetalSd   
...           ...        ...           ...       ...      ...         ...   
2598            6       1916          1950     Gable  CompShg     Wd Sdng   
2599            5       1955          1955       Hip  CompShg     BrkFace   
2600            6       1949          1950     Gable  CompShg     MetalSd   
2601            5       2000          2000     Gable  CompShg     VinylSd   
2602            5       1993          1994     Gable  CompShg     VinylSd   

     Exterior2nd MasVnrType  MasVnrArea ExterQual ExterCond Foundation  \
0        Wd Sdng       None         0.0        TA        TA     CBlock   
1        HdBoard    BrkFace       149.0        Gd        TA     CBlock   
2        HdBoard    BrkFace       149.0        Gd        TA     CBlock   
3   

In [167]:
housedf.to_csv('./data/Ames_HousePrice_processed')